marshal/consumer.go

```diff
@@ -458,76 +485,59 @@ func (c *Consumer) claimTopics() {
 	}
 }
 
-// updatesTopicClaims if changed. It should be noted that it expects c.lock.RLock
-// to be already acquired
-func (c *Consumer) safeUpdateTopicClaims(latestClaims map[string]bool, force bool) {
-	if !c.Terminated() {
-		c.lock.RLock()
-		defer c.lock.RUnlock()
-		// let's compare the new topic claims vs old ones. Upon change, we should trigger an update
-		// updateTopicClaims expects to be called with RLock held
-		c.updateTopicClaims(latestClaims, force)
+// sendTopicClaimUpdate can be called by various codepaths that have learned that there is
+// an update to send down to the users.
+func (c *Consumer) sendTopicClaimsUpdate() {
+	if c.Terminated() {
+		return
+	}
+
+	select {
+	case c.topicClaimsUpdated <- struct{}{}:
+		// Just sends a marker on the channel.
+	default:
 	}
 }
 
-// updatesTopicClaims if changed. It should be noted that it expects c.lock.RLock
-// to be already acquired
-func (c *Consumer) updateTopicClaims(latestClaims map[string]bool, force bool) {
-	changed := false
-	if force {
-		changed = true
-	} else {
-		// check for missing topic claims
-		for topic := range c.claimedTopics {
-			if !latestClaims[topic] {
-				changed = true
-				break
-			}
-		}
+// topicClaimsLoop sends an updated topic claim map every time we get a notification.
+func (c *Consumer) sendTopicClaimsLoop() {
+	defer close(c.topicClaimsChan)
 
-		if !changed {
-			// let's check if something in latest that is not in claimedTopics.
-			// checking for length only is not sufficient
-			for topic := range latestClaims {
-				if !c.claimedTopics[topic] {
-					changed = true
-					break
-				}
-			}
+	for range c.topicClaimsUpdated {
+		// Get consistent claims and send them
+		claims, err := c.GetCurrentTopicClaims()
+		if err != nil {
+			log.Errorf("Failed to send topic claims update: %s", err)
+			continue
 		}
-	}
-
-	if changed {
-		log.Infof("updating topic claims: %v\n", latestClaims)
-		c.lock.RUnlock()
-		defer c.lock.RLock()
-		func() {
-			// make sure that draining the channel and writing to it
-			// happens in the same atomic operation
-			c.lock.Lock()
-			defer c.lock.Unlock()
 
+		// Drain out any unconsumed updates
+	CONSUME:
+		for {
```

Comment:
You don't need this loop if you make the topicClaimsChan of size 0 and not 1. You're guaranteed to have at most one thing in the chan